<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkST0151ENSkillsNetwork20531532-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Regression Analysis**


Estimated time needed: **30** minutes


The goal of regression analysis is to describe the relationship between one set of variables called the dependent variables, and another set of variables, called independent or explanatory variables. When there is only one explanatory variable, it is called simple regression.


## Objectives


After completing this lab you will be able to:


*   Import Libraries
*   Regression analysis in place of the t-test
*   Regression analysis in place of ANOVA
*   Regression analysis in place of correlation


***


## Import Libraries


All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. If you run this notebook in a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.


In [ ]:
#install specific version of libraries used in lab
#! mamba install pandas==1.3.3
#! mamba install numpy=1.21.2
#! mamba install scipy=1.7.1-y
#!  mamba install seaborn=0.9.0-y
#!  mamba install matplotlib=3.4.3-y
#!  mamba install statsmodels=0.12.0-y

Import the libraries we need for the lab


In [1]:
import piplite
await piplite.install(['numpy'],['pandas'])

In [2]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

<ipython-input-2-bce3ea2f0fed>:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [40]:
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ST0151EN-SkillsNetwork/labs/teachingratings.csv'
resp = await fetch(URL)
ratings_url = io.BytesIO((await resp.arrayBuffer()).to_py())

Read in the csv file from the URL using the request library


In [41]:
ratings_df = pd.read_csv(ratings_url)
ratings_df.head(50)

minority  age  gender credits    beauty  eval division native tenure  \
0       yes   36  female    more  0.289916   4.3    upper    yes    yes   
1       yes   36  female    more  0.289916   3.7    upper    yes    yes   
2       yes   36  female    more  0.289916   3.6    upper    yes    yes   
3       yes   36  female    more  0.289916   4.4    upper    yes    yes   
4        no   59    male    more -0.737732   4.5    upper    yes    yes   
5        no   59    male    more -0.737732   4.0    upper    yes    yes   
6        no   59    male    more -0.737732   2.1    upper    yes    yes   
7        no   51    male    more -0.571984   3.7    upper    yes    yes   
8        no   51    male    more -0.571984   3.2    upper    yes    yes   
9        no   40  female    more -0.677963   4.3    upper    yes    yes   
10       no   40  female    more -0.677963   3.5    upper    yes    yes   
11       no   40  female    more -0.677963   4.1    upper    yes    yes   
12       no   40  female    more -0.677963   4.6    upper    yes    yes   
13       no   40  female    more -0.677963   3.8    upper    yes    yes   
14       no   40  female    more -0.677963   3.8    upper    yes    yes   
15       no   40  female    more -0.677963   3.8    upper    yes    yes   
16       no   40  female    more -0.677963   4.2    upper    yes    yes   
17       no   31  female    more  1.509794   4.4    upper    yes    yes   
18       no   31  female    more  1.509794   3.9    upper    yes    yes   
19       no   31  female    more  1.509794   4.5    upper    yes    yes   
20       no   31  female    more  1.509794   4.5    upper    yes    yes   
21       no   31  female    more  1.509794   4.4    upper    yes    yes   
22       no   31  female    more  1.509794   4.4    upper    yes    yes   
23       no   62    male    more  0.588569   4.2    upper    yes    yes   
24       no   62    male    more  0.588569   4.4    upper    yes    yes   
25       no   62    male    more  0.588569   4.5    upper    yes    yes   
26       no   62    male    more  0.588569   4.3    upper    yes    yes   
27       no   62    male    more  0.588569   4.6    upper    yes    yes   
28       no   62    male    more  0.588569   4.8    upper    yes    yes   
29       no   62    male    more  0.588569   4.3    upper    yes    yes   
30       no   33  female    more -0.126001   4.0    upper    yes    yes   
31       no   33  female    more -0.126001   4.0    upper    yes    yes   
32       no   33  female    more -0.126001   4.0    upper    yes    yes   
33       no   33  female    more -0.126001   3.9    upper    yes    yes   
34       no   33  female    more -0.126001   3.3    upper    yes    yes   
35       no   51  female    more -0.258190   3.4    upper    yes    yes   
36       no   51  female    more -0.258190   4.5    upper    yes    yes   
37       no   51  female    more -0.258190   4.5    upper    yes    yes   
38       no   51  female    more -0.258190   4.5    upper    yes    yes   
39       no   51  female    more -0.258190   2.5    lower    yes    yes   
40       no   51  female    more -0.258190   4.3    upper    yes    yes   
41       no   51  female    more -0.258190   4.5    upper    yes    yes   
42       no   33  female    more  0.149693   4.5    upper    yes    yes   
43       no   33  female    more  0.149693   4.7    upper    yes    yes   
44       no   33  female    more  0.149693   3.5    upper    yes    yes   
45       no   33  female    more  0.149693   4.5    upper    yes    yes   
46       no   33  female    more  0.149693   3.8    upper    yes    yes   
47       no   33  female    more  0.149693   4.5    upper    yes    yes   
48       no   33  female    more  0.149693   4.3    upper    yes    yes   
49       no   47    male    more  0.540917   3.9    upper    yes     no   

    students  allstudents  prof  PrimaryLast  vismin  female  single_credit  \
0         24           43     1            0       1       1              0   
1         86       

## Lab Exercises


In this section, you will learn how to run regression analysis in place of the t-test, ANOVA, and correlation


### Regression with T-test: Using the teachers rating data set, does gender affect teaching evaluation rates?


Initially, we had used the t-test to test if there was a statistical difference in evaluations for males and females, we are now going to use regression. We will state the null hypothesis:

*   $H\_0: β1$ = 0 (Gender has no effect on teaching evaluation scores)
*   $H\_1: β1$ is not equal to 0 (Gender has an effect on teaching evaluation scores)


We will use the female variable. female = 1 and male = 0


In [11]:
## X is the input variables (or independent variables)
X = ratings_df['female']
## y is the target/dependent variable
y = ratings_df['eval']
## add an intercept (beta_0) to our model
X = sm.add_constant(X)

model = sm.OLS(y, X).fit()
predictions = model.predict(X)
# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   eval   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.020
Method:                 Least Squares   F-statistic:                     10.56
Date:                Wed, 06 Nov 2024   Prob (F-statistic):            0.00124
Time:                        19:30:34   Log-Likelihood:                -378.50
No. Observations:                 463   AIC:                             761.0
Df Residuals:                     461   BIC:                             769.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.0690      0.034    121.288      0.000       4.003       4.135
female        -0.1680      0.052     -3.250      0.001      -0.270      -0.066
==============================================================================
Omnibus:                       17.625   Durbin-Watson:                   1.209
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               18.970
Skew:                          -0.496   Prob(JB):                     7.60e-05
Kurtosis:                       2.981   Cond. No.                         2.47
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Conclusion:** Like the t-test, the p-value is less than the alpha (α) level = 0.05, so we reject the null hypothesis as there is evidence that there is a difference in mean evaluation scores based on gender. The coefficient -0.1680 means that females get 0.168 scores less than men.


### Regression with ANOVA: Using the teachers' rating data set, does beauty  score for instructors  differ by age?


State the Hypothesis:

*   $H\_0: µ1 = µ2 = µ3$ (the three population means are equal)
*   $H\_1:$ At least one of the means differ


Then we group the data like we did with ANOVA


In [23]:
ratings_df.loc[(ratings_df['age'] <= 40), 'age_group'] = '40 years and younger'
ratings_df.loc[(ratings_df['age'] > 40)&(ratings_df['age'] < 57), 'age_group'] = 'between 40 and 57 years'
ratings_df.loc[(ratings_df['age'] >= 57), 'age_group'] = '57 years and older'

Use OLS function from the statsmodel library


In [24]:
from statsmodels.formula.api import ols
lm = ols('beauty ~ age_group', data = ratings_df).fit()
table= sm.stats.anova_lm(lm)
print(table)

              df      sum_sq    mean_sq          F        PR(>F)
age_group    2.0   20.422744  10.211372  17.597559  4.322549e-08
Residual   460.0  266.925153   0.580272        NaN           NaN


**Conclusion:** We can also see the same values for ANOVA like before and we will reject the null hypothesis since the p-value is less than 0.05 there is significant evidence that at least one of the means differ.


### Regression with ANOVA option 2


Create dummy variables - A dummy variable is a numeric variable that represents categorical data, such as gender, race, etc. Dummy variables are dichotomous, i.e they can take on only two quantitative values.


In [32]:
X = pd.get_dummies(ratings_df[['age_group']],  dtype = int)

     age_group_40 years and younger  age_group_57 years and older  \
0                                 1                             0   
1                                 1                             0   
2                                 1                             0   
3                                 1                             0   
4                                 0                             1   
..                              ...                           ...   
458                               1                             0   
459                               0                             1   
460                               0                             0   
461                               1                             0   
462                               0                             0   

     age_group_between 40 and 57 years  
0                                    0  
1                                    0  
2                                    0  
3      

In [31]:
y = np.asarray(ratings_df['beauty'])
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.071
Model:                            OLS   Adj. R-squared:                  0.065
Method:                 Least Squares   F-statistic:                     11.71
Date:                Wed, 06 Nov 2024   Prob (F-statistic):           2.10e-07
Time:                        20:05:31   Log-Likelihood:                -529.47
No. Observations:                 463   AIC:                             1067.
Df Residuals:                     459   BIC:                             1083.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================================
                                        coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
const                                 0.0138      0.028      0.495      0.621      -0.041       0.069
age_group_40 years and younger        0.3224      0.058      5.568      0.000       0.209       0.436
age_group_57 years and older         -0.2596      0.056     -4.616      0.000      -0.370      -0.149
age_group_between 40 and 57 years    -0.0489      0.045     -1.080      0.281      -0.138       0.040
==============================================================================
Omnibus:                       11.586   Durbin-Watson:                   0.434
Prob(Omnibus):                  0.003   Jarque-Bera (JB):               12.114
Skew:                           0.394   Prob(JB):                      0.00234
Kurtosis:                       2.913   Cond. No.                     1.12e+15
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 5.11e-28. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

You will get the same results and conclusion


### Correlation: Using the teachers' rating dataset, Is teaching evaluation score correlated with beauty score?


In [ ]:
## X is the input variables (or independent variables)
X = ratings_df['beauty']
## y is the target/dependent variable
y = ratings_df['eval']
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

**Conclusion:** p < 0.05 there is evidence of correlation between beauty and evaluation scores


## Practice Questions


### Question 1: Using the teachers' rating data set, does tenure affect beauty scores?

*   Use α = 0.05


In [42]:
## X is the input variables (or independent variables)
X = ratings_df['tenured_prof']
## y is the target/dependent variable
y = ratings_df['beauty']
## add an intercept (beta_0) to our model
X = sm.add_constant(X)

model = sm.OLS(y, X).fit()
predictions = model.predict(X)
# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 beauty   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.1689
Date:                Wed, 06 Nov 2024   Prob (F-statistic):              0.681
Time:                        20:15:00   Log-Likelihood:                -546.45
No. Observations:                 463   AIC:                             1097.
Df Residuals:                     461   BIC:                             1105.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            0.0284      0.078      0.363      0.717      -0.125       0.182
tenured_prof    -0.0364      0.089     -0.411      0.681      -0.210       0.138
==============================================================================
Omnibus:                       23.184   Durbin-Watson:                   0.461
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               23.229
Skew:                           0.507   Prob(JB):                     9.03e-06
Kurtosis:                       2.583   Cond. No.                         4.05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Double-click **here** for a hint.

<!-- The hint is below:
## State Hypothesis
Null Hypothesis: Mean beauty scores for tenured and non-tenured instructors are equal
Alternative Hypothesis: There is a difference in mean beauty scores for tenured and non-tenured instructors

## use the dummy variable for tenure - the OLS library doesn't recognize texts
X = ratings_df['tenured_prof']
-->


Double-click **here** for the solution.

<!-- The answer is below:
## put beauty scores in a list
y = ratings_df['beauty']
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()
p-value is greater than 0.05, so we fail to reject the null hypothesis as there is no evidence that the mean difference of tenured and untenured instructors are different
-->


### Question 2: Using the teachers' rating data set, does being an English speaker affect the number of students assigned to professors?

*   Use "allstudents"
*   Use α = 0.05 and α = 0.1


In [43]:
## X is the input variables (or independent variables)
X = ratings_df['English_speaker']
## y is the target/dependent variable
y = ratings_df['allstudents']
## add an intercept (beta_0) to our model
X = sm.add_constant(X)

model = sm.OLS(y, X).fit()
predictions = model.predict(X)
# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            allstudents   R-squared:                       0.007
Model:                            OLS   Adj. R-squared:                  0.005
Method:                 Least Squares   F-statistic:                     3.476
Date:                Wed, 06 Nov 2024   Prob (F-statistic):             0.0629
Time:                        20:18:40   Log-Likelihood:                -2654.2
No. Observations:                 463   AIC:                             5312.
Df Residuals:                     461   BIC:                             5321.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              29.6071     14.150      2.092      0.037       1.802      57.413
English_speaker    27.2158     14.598      1.864      0.063      -1.471      55.902
==============================================================================
Omnibus:                      429.792   Durbin-Watson:                   0.708
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            10527.126
Skew:                           4.129   Prob(JB):                         0.00
Kurtosis:                      24.852   Cond. No.                         8.01
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Double-click **here** for a hint.

<!-- The hint is below:
## State Hypothesis
Null Hypothesis: Mean number of students assigned to native English speakers vs non-native English speakers are equal
Alternative Hypothesis: There is a difference in mean number of students assigned to native English speakers vs non-native English speakers

## Is the instructor a native English speaker - make sure to use the binary variable "English speaker"
X = ratings_df['English_speaker']
## You can use the students or all students variable
y = ratings_df['allstudents']
-->


Double-click **here** for the solution.

<!-- The answer is below:
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

At α = 0.05, p-value is greater, we fail to reject the null hypothesis as there is no evidence that being a native English speaker or a non-native English speaker affects the number of students assigned to an instructor.
At α = 0.1, p-value is less, we reject the null hypothesis as there is evidence that there is a significant difference of mean number of students assigned to native English speakers vs non-native English speakers.
-->


### Question 3: Using the teachers' rating data set, what is the correlation between the number of students who participated in the evaluation survey and evaluation scores?

*   Use "students" variable


In [44]:
## X is the input variables (or independent variables)
X = ratings_df['students']
## y is the target/dependent variable
y = ratings_df["eval"]
## add an intercept (beta_0) to our model
X = sm.add_constant(X)

model = sm.OLS(y, X).fit()
predictions = model.predict(X)
# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   eval   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.5806
Date:                Wed, 06 Nov 2024   Prob (F-statistic):              0.446
Time:                        20:20:46   Log-Likelihood:                -383.46
No. Observations:                 463   AIC:                             770.9
Df Residuals:                     461   BIC:                             779.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.9823      0.033    119.689      0.000       3.917       4.048
students       0.0004      0.001      0.762      0.446      -0.001       0.002
==============================================================================
Omnibus:                       15.259   Durbin-Watson:                   1.198
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               16.283
Skew:                          -0.456   Prob(JB):                     0.000291
Kurtosis:                       2.888   Cond. No.                         74.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Double-click **here** for a hint.

<!-- The hint is below:
## create a list of students and evaluation socres
X = ratings_df['students']
y = ratings_df['eval']
-->


Double-click **here** for the solution.

<!-- The rest of the answer is below:
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

R-square is 0.001, R will be √0.001, correlation coefficient is 0.03 (close to 0). There is a very weak correlation between the number of students who participated in the evaluation survey and evaluation scores
-->


## Authors

[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)
<!--
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|
-->

<hr>

## <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
